packages/api/src/payments/payrexxSubscriptionPaymentProvider.ts

```diff
@@ -0,0 +1,449 @@
+      if (!isPayrexxExt) {
+        throw Error("It's not supported to reactivate a deactivated Payrexx subscription!")
+      }
+
+      // Disable Subscription (deactivation.upsert is defined if a deactivation is added)
+      if (deactivation.upsert || !subscription.autoRenew) {
+        await this.cancelRemoteSubscription(parseInt(isPayrexxExt.value, 10))
+
+        // Rewrite properties
+        isPayrexxExt.key = 'deactivated_payrexx_external_id'
+
+        // Rewrite subscription
+        subscription.autoRenew = false
+        subscription.deactivation = {
+          upsert: {
+            create: {
+              date: new Date(),
+              reason: SubscriptionDeactivationReason.userSelfDeactivated
+            },
+            update: {
+              date: new Date(),
+              reason: SubscriptionDeactivationReason.userSelfDeactivated
+            }
+          }
+        }
+
+        // Update subscription
+      } else {
+        const amount = subscription.monthlyAmount * getMonths(subscription.paymentPeriodicity)
+        await this.updateRemoteSubscription(parseInt(isPayrexxExt.value, 10), amount.toString())
+      }
+
+      // Return in the end
+      return next(params)
+    }
+    prisma.$use(hook())
+  }
+
+  async updatePaymentWithIntentState({
+    intentState,
+    paymentClient,
+    paymentsByID,
+    invoicesByID,
+    subscriptionClient,
+    userClient,
+    invoiceClient,
+    subscriptionPeriodClient,
+    invoiceItemClient
+  }: UpdatePaymentWithIntentStateProps): Promise<any> {
+    const apiData = JSON.parse(intentState.paymentData ? intentState.paymentData : '{}')
+    const rawSubscription = apiData.subscription
+    const subscriptionId = rawSubscription.id
+
+    if (intentState.state === PaymentState.paid) {
+      const subscriptionValidUntil = startOfDay(parseISO(rawSubscription.valid_until))
+
+      // Get subscription
+      const subscription = await findSubscriptionByExternalId(subscriptionClient, subscriptionId)
+      if (!subscription) {
+        logger('payrexxSubscriptionPaymentProvider').warn(
+          `Subscription ${subscriptionId} received from payrexx webhook not found!`
+        )
+        return
+      }
+
+      // Calculate max possible Extension length for subscription security margin of 7 days
+      const maxSubscriptionExtensionLength = sub(subscriptionValidUntil, {days: 7})
+
+      // Find last paid period in array
+      let longestPeriod
+      for (const period of subscription.periods) {
+        if (period.Invoice.paidAt && (!longestPeriod || period.endsAt > longestPeriod.endsAt)) {
+          longestPeriod = period
+        }
+      }
+
+      // If no period is found throw error
+      if (!longestPeriod) throw Error(`No period found in subscription ${subscriptionId}`)
+
+      // Skip if subscription is already renewed
+      if (maxSubscriptionExtensionLength <= startOfDay(longestPeriod.endsAt)) {
```

Comment:
"maxSubscriptionExtensionLength" is not a length but a date. That is misleading, when reading the code.